### PR TITLE
Add some basic types necessary for HTTP/2

### PR DESCRIPTION
**Pull request by bryce-anderson**
Just primitives and helpers.

### CODE REVIEW

File: http/src/main/scala/org/http4s/blaze/http/http2/Http2Exception.scala [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        val stream = ex match {
+          case ex: Http2StreamException => Some(ex.stream)
+          case _ => None
+        }
+        Some(stream -> ex.getMessage)
+      }
+      else None
+    }
+
+    def unapply(code: Int): Option[Unit] = {
```
Comment (rossabaker) [inline http/src/main/scala/org/http4s/blaze/http/http2/Http2Exception.scala]: Minor, maybe not wanted, but don't forget `unapply` can return `Boolean`.

File: http/src/main/scala/org/http4s/blaze/http/http2/Http2Result.scala [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package org.http4s.blaze.http.http2
+
+/** Result type of many of the codec methods */
+sealed trait Http2Result
```
Comment (rossabaker) [inline http/src/main/scala/org/http4s/blaze/http/http2/Http2Result.scala]: These are probably not ever in a collection, but extending `Product with Serializable` in the trait improves inference.  I've tried to get into that habit everywhere.

File: http/src/main/scala/org/http4s/blaze/http/http2/Http2Settings.scala [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+  val HEADER_TABLE_SIZE      = makeKey(0x1, "SETTINGS_HEADER_TABLE_SIZE")
+  val ENABLE_PUSH            = makeKey(0x2, "SETTINGS_ENABLE_PUSH")
+  val MAX_CONCURRENT_STREAMS = makeKey(0x3, "SETTINGS_MAX_CONCURRENT_STREAMS")
+  val INITIAL_WINDOW_SIZE    = makeKey(0x4, "SETTINGS_INITIAL_WINDOW_SIZE")
+  val MAX_FRAME_SIZE         = makeKey(0x5, "SETTINGS_MAX_FRAME_SIZE")
+  val MAX_HEADER_LIST_SIZE   = makeKey(0x6, "SETTINGS_MAX_HEADER_LIST_SIZE")
+}
+
+/** Internal mutable representation of the [[Http2Settings]] */
+private final case class MutableHttp2Settings private(
```
Comment (rossabaker) [inline http/src/main/scala/org/http4s/blaze/http/http2/Http2Settings.scala]: What are we getting out of being a case class if it's mutable?

File: http/src/main/scala/org/http4s/blaze/http/http2/bits.scala [hunk 4]
```diff
@@ -0,0 +1,64 @@
+package org.http4s.blaze.http.http2
+
+import java.nio.ByteBuffer
+import java.nio.charset.StandardCharsets
+
+object bits {
```
Comment (rossabaker) [inline http/src/main/scala/org/http4s/blaze/http/http2/bits.scala]: Private so you can change it later?

File: http/src/main/scala/org/http4s/blaze/http/http2/Http2Exception.scala (hunk 1 above)
Comment (bryce-anderson) [reply http/src/main/scala/org/http4s/blaze/http/http2/Http2Exception.scala]: TIL.

File: http/src/main/scala/org/http4s/blaze/http/http2/Http2Settings.scala (hunk 3 above)
Comment (bryce-anderson) [reply http/src/main/scala/org/http4s/blaze/http/http2/Http2Settings.scala]: Probably initially a good `toString`, but nothing that I can think of now. I'll remove it.

File: http/src/main/scala/org/http4s/blaze/http/http2/bits.scala (hunk 4 above)
Comment (bryce-anderson) [reply http/src/main/scala/org/http4s/blaze/http/http2/bits.scala]: Good call.

**Pull request merged**